### PR TITLE
Basic Android Instrumentation Test

### PR DESCRIPTION
**Pull request by yschimke**
Run separately against release version currently.  i.e. not to run inside CI.

Mainly useful to define a set of tests we expect to run against real emulators or phones to check for regressions or confirm reported bugs.

Specifically:
 - Tests don't run against a mocked environment like robolectric.  Only emulator or phone.
 - Tests are not part of regular CI.
 - Tests are currently against a released version, but could change to master if that is easy.
 - Tests are a separate project in same repo.

Run with

```
$ ./gradlew :android-test:connectedCheck
```

### CODE REVIEW

File: android-test/app/src/main/res/values-w820dp/dimens.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<resources>
+    <!-- Example customization of dimensions originally defined in res/values/dimens.xml
+         (such as screen margins) for screens with more than 820dp of available width. This
+         would include 7" and 10" devices in landscape (~960dp and ~1280dp respectively). -->
+    <dimen name="activity_horizontal_margin">64dp</dimen>
```
Comment (JakeWharton) [inline android-test/app/src/main/res/values-w820dp/dimens.xml]: Almost all of these resource files are not necessary.

File: android-test/build.gradle [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+allprojects {
+    repositories {
+        jcenter()
+        google()
+    }
+}
+
+ext {
+    kotlinVersion = "1.3.40"
```
Comment (JakeWharton) [inline android-test/build.gradle]: Don't like that all of these are separate from the other variables in the enclosing project

**Comment by yschimke** [comment]
@JakeWharton Thanks for the review.  I took one of the google example instrumentation tests, I didn't want to spend too much time cleaning it up before getting feedback from others on whether it was desirable.  Based on feedback I'll clean it up and land when I have some time.  Thanks!

File: settings.gradle [hunk 3]
```diff
@@ -1,4 +1,11 @@
+rootProject.name = 'okhttp-parent'
```
Comment (yschimke) [inline settings.gradle]:
```
> Failed to apply plugin [id 'com.android.library']
   > Your project contains 2 or more modules with the same identification com.squareup.okhttp3:okhttp
     at ":" and ":okhttp".
     You must use different identification (either name or group) for each modules.
```

File: settings.gradle [hunk 4]
```diff
@@ -1,4 +1,11 @@
...
+
 include ':mockwebserver'
+if (!properties.containsKey('android.injected.invoked.from.ide') && System.getenv(
```
Comment (yschimke) [inline settings.gradle]: @swankjesse @JakeWharton  This should make the code pretty safe for both normal development and also CI.  But allow us to do real local tests against phones etc to flush out bugs.

Thoughts?

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 5]
```diff
@@ -0,0 +1,86 @@
+package okhttp.android.test
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: nit: copyright

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    response.use {
+      assertEquals(Protocol.HTTP_2, response.protocol)
+      if (Build.VERSION.SDK_INT >= 29) {
+        assertEquals(TlsVersion.TLS_1_3, response.handshake?.tlsVersion)
+      } else {
+        assertEquals(TlsVersion.TLS_1_2, response.handshake?.tlsVersion)
+      }
+      assertEquals(200, response.code)
+      assertEquals("com.android.org.conscrypt.Java8FileDescriptorSocket", socketClass)
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Not in love with this. Testing behavior above is sufficient?

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  @Test
+  fun testHttpRequestBlocked() {
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: This won’t work on all Androids

File: settings.gradle (hunk 4 above)
Comment (swankjesse) [reply settings.gradle]: Works for me. Got an issue to link to so we know when we can delete this?

**Comment by swankjesse** [comment]
Also! Tests that call real hosts doesn’t seem like a very reliable strategy.

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt (hunk 6 above)
Comment (yschimke) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: I specifically want to confirm that we are using an Android socket here.  So I'd like one test that confirms this.

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt (hunk 7 above)
Comment (yschimke) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: I'll add junit assumptions

**Comment by yschimke** [comment]
> Mixed feels on this.
> 
> When I was playing with this I wrote an Android app that ran the standard OkHttp test suite, with a custom suppression list. It was hackier and uglier but it meant I could pretty easily figure out where the problems were.

I'll try to make sure it stays out of the way, but TBH it was a godsend for testing out the socket changes etc.  So much quicker.  Benefits overall

1) It's the quickest target for someone to add a bug reproduction to, without generating a new project and uploading to github.
2) It's good for us developing anything internally that needs to run on Android.

If I keep this out of the way, give me some lee-way and I'll land close to as-is, then add more MockWebserver tests over time to beef up the capabilities.  It should give us better coverage of not breaking android with changes sooner.  It becomes trivial to run these tests locally and frequently against a real phone etc.

**Comment by yschimke** [comment]
> Also! Tests that call real hosts doesn’t seem like a very reliable strategy.

I think we need a couple but I can put them behind an assumeNetwork() check.

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt (hunk 6 above)
Comment (swankjesse) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Sounds good!

**Pull request merged**